File: src/main/java/net/benwoodworth/fastcraft/core/api/FastCraftApi.java
package net.benwoodworth.fastcraft.core.api;

import net.benwoodworth.fastcraft.core.api.dependencies.config.FcConfigProvider;
import net.benwoodworth.fastcraft.core.api.dependencies.inventory.FcInventoryProvider;
import net.benwoodworth.fastcraft.core.api.dependencies.permissions.FcPermissionService;
import net.benwoodworth.fastcraft.core.api.dependencies.player.FcPlayerProvider;
import net.benwoodworth.fastcraft.core.api.dependencies.recipes.FcRecipeService;
import net.benwoodworth.fastcraft.core.api.gui.GuiApi;
import org.jetbrains.annotations.NotNull;

/**
 * An API other plugins can use to interface with FastCraft+.
 *
 * @param <TItem> The native item type.
 */
public class FastCraftApi<TItem> {
    @NotNull
    private final FcInventoryProvider<TItem> inventoryProvider;
    @NotNull
    private final FcPlayerProvider<TItem> playerProvider;
    @NotNull
    private final FcRecipeService<TItem> recipeService;
    @NotNull
    private final FcPermissionService permissionService;
    @NotNull
    private final FcConfigProvider configProvider;
    @NotNull
    private final GuiApi<TItem> guiApi;

    public FastCraftApi(@NotNull FcInventoryProvider<TItem> inventoryProvider,
                        @NotNull FcPlayerProvider<TItem> playerProvider,
                        @NotNull FcRecipeService<TItem> recipeService,
                        @NotNull FcPermissionService permissionService,
                        @NotNull FcConfigProvider configProvider) {

        this.inventoryProvider = inventoryProvider;
        this.playerProvider = playerProvider;
        this.recipeService = recipeService;
        this.permissionService = permissionService;
        this.configProvider = configProvider;

        guiApi = new GuiApi<>(this);
    }

    @NotNull
    public FcInventoryProvider<TItem> getInventoryProvider() {
        return inventoryProvider;
    }

    @NotNull
    public FcPlayerProvider<TItem> getPlayerProvider() {
        return playerProvider;
    }

    @NotNull
    public FcRecipeService<TItem> getRecipeService() {
        return recipeService;
    }

    @NotNull
    public FcPermissionService getPermissionService() {
        return permissionService;
    }

    @NotNull
    public FcConfigProvider getConfigProvider() {
        return configProvider;
    }

    @NotNull
    public GuiApi<TItem> getGuiApi() {
        return guiApi;
    }
}

File: src/main/kotlin/net/benwoodworth/fastcraft/core/api/FastCraftApi.kt
package net.benwoodworth.fastcraft.core.api

import net.benwoodworth.fastcraft.core.FastCraft
import net.benwoodworth.fastcraft.core.dependencies.inventory.FcItem
import javax.inject.Inject

/**
 * A FastCraft API for use by other plugins.
 *
 * @param TItem the item type
 */
class FastCraftApi<TItem : FcItem<*>> @Inject constructor(
        /** The instance of FastCraft. */
        val fastCraft: FastCraft<TItem>
) {

//    /** TODO
//     * Checks to see if an inventory is a GUI.
//     *
//     * @param inventory the inventory to check
//     * @return `true` if the inventory is a GUI
//     */
//    fun isGuiInventory(inventory: TInventory): Boolean {
//        val fcInv = fastCraft.dependencies.inventoryProvider.adaptInventory(inventory)
//        val carrier = fcInv.carrier
//        return carrier is Gui<*, *>
//    }
}
